### PR TITLE
V prediction ddim

### PR DESCRIPTION
**Pull request by bglick13**
changes to DDIM scheduler to allow for v prediction training. Accompanying example script to train on butterflies dataset

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 1]
```diff
@@ -27,12 +27,24 @@
...
+    return out
+
+
 @dataclass
 # Copied from diffusers.schedulers.scheduling_ddpm.DDPMSchedulerOutput with DDPM->DDIM
 class DDIMSchedulerOutput(BaseOutput):
     """
     Output class for the scheduler's step function output.
 
+
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddim.py:47]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 2]
```diff
@@ -145,51 +171,70 @@ def __init__(
...
 
     def scale_model_input(self, sample: torch.FloatTensor, timestep: Optional[int] = None) -> torch.FloatTensor:
         """
         Ensures interchangeability with schedulers that need to scale the denoising model input depending on the
         current timestep.
 
+
         Args:
             sample (`torch.FloatTensor`): input sample
             timestep (`int`, optional): current timestep
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddim.py:204]:
```suggestion
            timestep (`int`, optional): current timestep
            
```

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 3]
```diff
@@ -145,51 +171,70 @@ def __init__(
...
+            denominator = (1 - self.alphas[timestep] ** 2 / alpha_prev**2).clamp(min=1.0e-7).sqrt()
+            ddim_sigma = (numerator * denominator).clamp(min=1.0e-7)
+            variance = (sigma_prev**2 - ddim_sigma**2).clamp(min=1.0e-7).sqrt()
         return variance
 
     def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.device] = None):
         """
         Sets the discrete timesteps used for the diffusion chain. Supporting function to be run before inference.
 
+
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddim.py:237]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 4]
```diff
@@ -218,6 +262,7 @@ def step(
         Predict the sample at the previous timestep by reversing the SDE. Core function to propagate the diffusion
         process from the learned model outputs (most often the predicted noise).
 
+
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddim.py:265]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 5]
```diff
@@ -145,51 +171,70 @@ def __init__(
...
         beta_prod_t_prev = 1 - alpha_prod_t_prev
 
-        variance = (beta_prod_t_prev / beta_prod_t) * (1 - alpha_prod_t / alpha_prod_t_prev)
-
+        if self.variance_type == "fixed":
+            variance = (beta_prod_t_prev / beta_prod_t) * (1 - alpha_prod_t / alpha_prod_t_prev)
+        elif self.variance_type == "v_diffusion":
+            # If eta > 0, adjust the scaling factor for the predicted noise
+            # downward according to the amount of additional noise to add
+            # variance = torch.log(self.betas[timestep] * (1 - alpha_prod_t_prev) / (1 - alpha_prod_t))
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddim.py:221]:
```suggestion
```
